src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.GeneratedFileReplacingCompilationTracker.cs

```diff
@@ -0,0 +1,217 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using System.Threading.Tasks;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal partial class SolutionState
+    {
+        /// <summary>
+        /// An implementation of <see cref="ICompilationTracker"/> that takes a compilation from another compilation tracker and updates it
+        /// to return a generated document with a specific content, regardless of what the generator actually produces. In other words, it says
+        /// "take the compilation this other thing produced, and pretend the generator gave this content, even if it wouldn't."
```

Comment:
i'm confsued about when this would happen.  can you give an example in teh comments why we need this.  Why do we need to pretend anything about the generator.  Why can't we use it as the source of truth?  I thought generated files were readonly, so it's not like the user can edit them.